source/tutorials/file-sets.md

```diff
@@ -0,0 +1,725 @@
+It abstracts over built-in functionality and offers a safer and more convenient interface.
+
+## File sets
+
+The file set library is based on the concept of _file sets_,
+a data type representing a collection of local files.
+File sets can be created, composed, and manipulated with the various functions of the library.
+
+The easiest way to experiment with the library is to use it through [`nix repl`](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-repl):
+
+```shell-session
+$ nix repl -f channel:nixos-unstable
+...
+nix-repl> fs = lib.fileset
+```
+
+The [`trace`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.trace) function pretty-prints the files included in a given file set:
+
+```shell-session
+nix-repl> fs.trace ./. null
+trace: /home/user (all files in directory)
+null
+```
+
+All functions that expect a file set for an argument also accept a [path](https://nixos.org/manual/nix/stable/language/values#type-path).
+Such path arguments are then [implicitly coerced](https://nixos.org/manual/nixpkgs/unstable/#sec-fileset-path-coercion), and the resulting file sets contain _all_ files under the given path.
+In the previous trace this is indicated by `(all files in directory)`.
+
+:::{tip}
+The `trace` function pretty-prints its first agument and returns its second argument.
+But since you often just need the pretty-printing in `nix repl`, you can omit the second argument:
+
+```shell-session
+nix-repl> fs.trace ./.
+trace: /home/user (all files in directory)
+«lambda @ /nix/store/1czr278x24s3bl6qdnifpvm5z03wfi2p-nixpkgs-src/lib/fileset/default.nix:555:8»
+```
+:::
+
+Even though file sets conceptually contain local files, these files are *never* added to the Nix store unless explicitly requested.
+You don't have to worry about accidentally copying secrets into the world-readable store.
+
+In this example, although we pretty-printed the home directory, no files were copied.
+This is also evident from the expression evaluating instantly.
+
+:::{note}
+This is in contrast to coercion of paths to strings such as in `"${./.}"`,
+which copies the whole directory to the Nix store on evaluation!
+:::
+
+:::{warning}
+With [experimental Flakes](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake), a local directory containing `flake.nix` is always copied into the Nix store *completely* unless it is a Git repository!
+:::
+
+This implicit coercion also works for files:
+
+```shell-session
+nix-repl> fs.trace /etc/nix/nix.conf
+trace: /etc/nix
+trace: - nix.conf (symlink)
+```
+
+In addition to the included file, this also prints its [file type](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-readFileType).
+
+If a given path doesn't exist, the library will complain:
+
+```shell-session
+nix-repl> fs.trace /etc/nix/nix.nix
+error: lib.fileset.trace: Argument (/etc/nix/nix.nix)
+  is a path that does not exist.
+```
```

Comment:
It's fair to make users aware of this, since there's a good chance they will introduce typos. It's not wrong to have it but I won't complain if it was removed.